File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionDaysBetween.java
package org.hisp.dhis.rules.functions;

import org.hisp.dhis.rules.models.TimeInterval;
import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
import org.hisp.dhis.rules.parser.expression.function.ScalarFunctionToEvaluate;
import org.joda.time.Days;

import static org.hisp.dhis.antlr.AntlrParserUtils.castDate;
import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;

public class RuleFunctionDaysBetween
    extends ScalarFunctionToEvaluate
{
    /**
     * Function which will return the number of days between the two given dates.
     *
     * @param start the start date.
     * @param end   the end date.
     * @return number of days between dates.
     */
    private Integer daysBetween( String start, String end )
    {

        TimeInterval interval = RuleFunction.getTimeInterval( start, end );

        if ( interval.isEmpty() )
        {
            return 0;
        }

        return Days.daysBetween( interval.getStartDate(), interval.getEndDate() ).getDays();
    }

    @Override
    public Object evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        return String.valueOf( daysBetween(
            visitor.castStringVisit( ctx.expr( 0 ) ),
            visitor.castStringVisit( ctx.expr( 1 ) ) ) );
    }

    @Override
    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        castDate( visitor.visit( ctx.expr( 0 ) ) );
        castDate( visitor.visit( ctx.expr( 1 ) ) );

        return CommonExpressionVisitor.DEFAULT_DATE_VALUE;
    }
}

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionDaysBetween.kt
package org.hisp.dhis.rules.functions

import org.hisp.dhis.rules.RuleVariableValue
import java.text.ParseException
import java.text.SimpleDateFormat

const val D2_DAYS_BETWEEN = "d2:daysBetween"

internal class RuleFunctionDaysBetween : RuleFunction() {
    override fun evaluate(arguments: List<String>,
                          valueMap: Map<String, RuleVariableValue>, supplementaryData: Map<String, List<String>>?): String {
        require(arguments.size == 2) {
            "Two arguments were expected, " +
                    arguments.size + " were supplied"
        }
        return daysBetween(arguments[0], arguments[1]).toString()
    }

    /**
     * Function which will return the number of days between the two given dates.
     *
     * @param start the start date.
     * @param end   the end date.
     * @return number of days between dates.
     */
    private fun daysBetween(start: String, end: String): Int {
        if (isEmpty(start) || isEmpty(end)) {
            return 0
        }
        val format = SimpleDateFormat()
        format.applyPattern(DATE_PATTERN)
        return try {
            val startDate = format.parse(start)
            val endDate = format.parse(end)
            java.lang.Long.valueOf((endDate.time - startDate.time) / 86400000).toInt()
        } catch (parseException: ParseException) {
            throw RuntimeException(parseException)
        }
    }
}